docs/enhancements/0002_day0-integration.md

```diff
@@ -0,0 +1,257 @@
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kmm-kmod-dockerfile
+data:
+  dockerfile: |
+
+    ARG DTK_AUTO
+    ARG OS_AUTO
+
+    FROM ${DTK_AUTO} as builder
+
+    ARG KERNEL_VERSION
+    ...
+    RUN KERNEL_SRC_DIR=/lib/modules/${KERNEL_VERSION}/build make all
+
+    FROM ${OS_AUTO}
+
+    ARG KERNEL_VERSION
+
+    COPY --from=builder /usr/src/kernel-module-management/ci/kmm-kmod/kmm_ci_a.ko /opt/lib/modules/${KERNEL_VERSION}/
+    COPY --from=builder /usr/src/kernel-module-management/ci/kmm-kmod/kmm_ci_b.ko /opt/lib/modules/${KERNEL_VERSION}/
+    RUN depmod -b /opt ${KERNEL_VERSION}
+```
+
+`OS_AUTO` will fetch the `rhel-coreos` image from the release-payload of the
+cluster.
+The DTK `ImageStream` on the cluster will contain the DTK image that fits the
+release-payload and not the running OS on the nodes (which may have a different
+kernel), therefore, `DTK_AUTO` should only be used if `OS_AUTO` is used.
+
+### Signing
+
+If a customer wishes KMM to re-build his kmods, they should add a `build` section
+in the `ModuleDay0` as follow
+```
+apiVersion: kmm.sigs.x-k8s.io/v1beta1
+kind: ModuleDay0
+spec:
+  osMapping:
+    literal: "Red Hat Enterprise Linux CoreOS 414.92.202312311229-0 (Plow)"
+    containerImage: quay.io/ybettan/rhcos:414.92.202312311229-0
+    sign:
+      unsignedImage: quay.io/ybettan/rhcos:414.92.202312311229-0-unsigned
+      certSecret:
+        name: kmm-kmod-signing-cert
+      keySecret:
+        name: kmm-kmod-signing-key
+      filesToSign:
+        - /opt/lib/modules/${KERNEL_FULL_VERSION}/kmm_ci_a.ko
+  machineConfigName: 99-ybettan-external-image
+```
+
+The sign will trigger if all the following conditions are met
+  * `quay.io/ybettan/rhcos:414.92.202312311229-0` is different from `spec.osImageURL` in the `MachineConfig`;
+  * `quay.io/ybettan/rhcos:414.92.202312311229-0-unsigned` exist (or a `build` section exists)
+  * A `sign` section exists in the `osMapping`
+
+Once signed and pushed, KMM will modify the `MachineConfig`'s `spec.osImageURL`
+with the new image.
+
+### Firmware
+
+No much to do here, the user can put his firmware files directly in
+`/lib/firmware` on the nodes in the custom container image or set
+`firmware_class.path` in their image's `sysctl.conf` so the firmware feature of
+KMM is a bit useless here.
+
+### Preflight
+
+In day2 we need preflight because we upgrade the cluster before building the
+artifacts but in day0 we first generate the artifacts and only then modify the
+`MachineConfig` with the new image, therefore, if a build/sign failed we won't
+reach the stage of modifying the `MachineConfig` at all and the cluster will
+remain valid.
+
+### Cluster upgrade
+
+When a `MachineConfig` uses `osImageURL`, it is called
+[image-layering](https://docs.openshift.com/container-platform/4.14/post_installation_configuration/coreos-layering.html).
+
+When image-layering is used, the OCP layer is being "detached" from the underlying
```

Comment:
is it still the case? I was under the assumption that it should be addressed in future releases

Reply:
Me too. I haven't checked its status recently. Will ask.

Reply:
It is still the same behavior indeed.
Looks like the [proposed-expriance](https://github.com/rh-ecosystem-edge/kernel-module-management/pull/957/files#diff-502e7d789feb94b276610d47444683bfd32820e1594f9413f0b310843f5682b4R238-R248) may duplicate some effort planned in https://issues.redhat.com/browse/MCO-665.

@romfreiman Do you think we should avoid upgrading the nodes after a cluster upgrade due to the recent info about planned action in the MCO space?

Reply:
i rather not have 2 mechanisms to achieve the same. And, it will introduce more reboots - 10m pare baremetal server.

Reply:
Then I will drop this "improvement" from the proposal.
FYI @qbarrand